File: runtime/csharp/Prompty.Core/TurnEngineRequest.cs
// Copyright (c) Microsoft. All rights reserved.

namespace Prompty.Core;

/// <summary>
/// Request accepted by the canonical <see cref="TurnEngine"/>. Mirrors the Rust reference's
/// <c>TurnEngineRequest</c> (runtime/rust/prompty/src/engine/turn.rs) field for field. A fresh
/// turn is built with <see cref="TurnEngineRequest(string, string, IList{Message})"/>; an
/// interrupted turn is resumed with <see cref="ResumeFrom"/>/<see cref="FromResume"/> and the
/// reconciliation-specific resume factories.
/// </summary>
public sealed class TurnEngineRequest
{
    public TurnEngineRequest(string sessionId, string turnId, IList<Message> messages)
    {
        SessionId = sessionId;
        TurnId = turnId;
        Messages = messages;
        StablePrefixMessages = messages.Count;
    }

    /// <summary>Session this turn belongs to.</summary>
    public string SessionId { get; set; }

    /// <summary>Turn identifier, unique within the session.</summary>
    public string TurnId { get; set; }

    /// <summary>Stable identifier of this engine run. Empty means the engine assigns one at run start.</summary>
    public string RunId { get; set; } = string.Empty;

    /// <summary>Run identifier of the parent run when this run was delegated.</summary>
    public string? ParentRunId { get; set; }

    /// <summary>Zero-based delegation nesting depth; 0 for a top-level run.</summary>
    public int DelegationDepth { get; set; }

    /// <summary>Conversation messages seen so far.</summary>
    public IList<Message> Messages { get; set; }

    /// <summary>Opaque turn inputs, echoed into context snapshots and host policy requests.</summary>
    public object? Inputs { get; set; }

    /// <summary>Maximum number of model/tool iterations permitted for this run.</summary>
    public int MaxIterations { get; set; } = 10;

    /// <summary>Maximum model invocation attempts per iteration before failing the turn.</summary>
    public int MaxModelAttempts { get; set; } = 3;

    /// <summary>Iteration to execute first. Non-zero values are used when resuming a checkpoint.</summary>
    public int StartIteration { get; set; }

    /// <summary>Last committed event sequence before this run.</summary>
    public long InitialSequence { get; set; }

    /// <summary>Number of leading messages the engine and host policy must not rewrite.</summary>
    public int StablePrefixMessages { get; set; }

    /// <summary>Whether the provider-side context can be reconstructed from <see cref="Messages"/> alone.</summary>
    public InvocationContextPortability Portability { get; set; } = InvocationContextPortability.Portable;

    /// <summary>Opaque references to delegated provider-side state, required when <see cref="Portability"/> is not portable.</summary>
    public IList<DelegatedStateReference> DelegatedState { get; set; } = [];

    /// <summary>Invocation id in progress when this request was built, if any.</summary>
    public string? ActiveInvocationId { get; set; }

    /// <summary>Tool requests from the most recent model response not yet executed.</summary>
    public IList<ModelToolRequest> PendingToolRequests { get; set; } = [];

    /// <summary>Tool results already committed for the in-flight model/tool exchange.</summary>
    public IList<ModelToolResult> CompletedToolResults { get; set; } = [];

    /// <summary>Number of model iterations completed so far.</summary>
    public int CompletedModelIterations { get; set; }

    /// <summary>Whether the turn is blocked pending an explicit reconciliation resolution.</summary>
    public bool ReconciliationRequired { get; set; }

    /// <summary>Durable reconciliation state for an indeterminate model invocation, if any.</summary>
    public ModelReconciliationState? ModelReconciliation { get; set; }

    /// <summary>Output pending final host policy application.</summary>
    public object? PendingOutput { get; set; }

    /// <summary>Whether a model response with no further tool requests is ready to commit.</summary>
    public bool FinalOutputReady { get; set; }

    /// <summary>Model response awaiting completion of its tool exchange.</summary>
    public ModelInvocationResponse? PendingModelResponse { get; set; }

    /// <summary>Whether host policy has already been applied for the current iteration (skip re-applying once).</summary>
    public bool PolicyAppliedForIteration { get; set; }

    /// <summary>Resolved tool result supplied when resuming after an indeterminate tool effect.</summary>
    public ModelToolResult? ReconciliationResolution { get; set; }

    /// <summary>Resolved model response supplied when resuming after an indeterminate model invocation.</summary>
    public ModelInvocationResponse? ModelReconciliationResolution { get; set; }

    /// <summary>Set the stable run identifier for this run. An empty value lets the engine assign one at run start.</summary>
    public TurnEngineRequest WithRunId(string runId)
    {
        RunId = runId;
        return this;
    }

    /// <summary>
    /// Mark this run as delegated from a parent run, carrying the parent run identifier and
    /// nesting one level deeper than the parent.
    /// </summary>
    public TurnEngineRequest DelegatedUnder(string parentRunId, int parentDelegationDepth)
    {
        ParentRunId = parentRunId;
        DelegationDepth = parentDelegationDepth + 1;
        return this;
    }

    /// <summary>Resume while continuing a journal whose tail may follow the checkpoint.</summary>
    public static TurnEngineRequest ResumeFrom(EngineCheckpoint checkpoint, int maxIterations, long lastJournalSequence)
    {
        var noPendingWork = (checkpoint.PendingToolRequests is null || checkpoint.PendingToolRequests.Count == 0)
            && checkpoint.PendingModelResponse is null
            && !checkpoint.FinalOutputReady
            && !checkpoint.ReconciliationRequired;
        var startIteration = checkpoint.ResumeSameIteration
            ? checkpoint.Iteration
            : noPendingWork ? checkpoint.Iteration + 1 : checkpoint.Iteration;

        return new TurnEngineRequest(checkpoint.SessionId, checkpoint.TurnId, [.. checkpoint.Messages])
        {
            RunId = checkpoint.RunId,
            ParentRunId = checkpoint.ParentRunId,
            DelegationDepth = checkpoint.DelegationDepth,
            StablePrefixMessages = checkpoint.StablePrefixMessages,
            Inputs = checkpoint.Inputs,
            MaxIterations = maxIterations,
            MaxModelAttempts = 3,
            StartIteration = startIteration,
            InitialSequence = Math.Max(lastJournalSequence, checkpoint.LastSequence),
            Portability = checkpoint.ContextState.Portability,
            DelegatedState = checkpoint.ContextState.DelegatedState is null ? [] : [.. checkpoint.ContextState.DelegatedState],
            ActiveInvocationId = checkpoint.ActiveInvocationId,
            PendingToolRequests = checkpoint.PendingToolRequests is null ? [] : [.. checkpoint.PendingToolRequests],
            CompletedToolResults = checkpoint.CompletedToolResults is null ? [] : [.. checkpoint.CompletedToolResults],
            CompletedModelIterations = checkpoint.CompletedModelIterations,
            ReconciliationRequired = checkpoint.ReconciliationRequired,
            ModelReconciliation = checkpoint.ModelReconciliation,
            PendingOutput = checkpoint.PendingOutput,
            FinalOutputReady = checkpoint.FinalOutputReady,
            PendingModelResponse = checkpoint.PendingModelResponse,
            PolicyAppliedForIteration = checkpoint.PolicyAppliedForIteration,
            ReconciliationResolution = null,
            ModelReconciliationResolution = null,
        };
    }

    /// <summary>
    /// Resume after the host resolves an indeterminate tool effect. Patches the checkpoint's
    /// completed tool result and, if the conversation batch was already flushed, the matching
    /// tool-result message, then resumes from the patched checkpoint.
    /// </summary>
    /// <exception cref="TurnEngineInvalidRequestException">
    /// The checkpoint does not require tool reconciliation, the resolved result is still
    /// indeterminate, or the checkpoint has no matching indeterminate tool request.
    /// </exception>
    public static TurnEngineRequest ResumeAfterReconciliation(
        EngineCheckpoint checkpoint,
        int maxIterations,
        long lastJournalSequence,
        ModelToolResult resolvedResult)
    {
        if (!checkpoint.ReconciliationRequired)
        {
            throw new TurnEngineInvalidRequestException("checkpoint does not require reconciliation");
        }

        if (checkpoint.ModelReconciliation is not null)
        {
            throw new TurnEngineInvalidRequestException(
                "checkpoint requires model reconciliation, not tool reconciliation");
        }

        if (resolvedResult.Outcome == ModelToolOutcome.Indeterminate)
        {
            throw new TurnEngineInvalidRequestException("resolved tool result must have a determinate outcome");
        }

        var resolved = CloneForPatch(checkpoint);
        var results = resolved.CompletedToolResults ??= [];
        var index = IndexOfToolResult(results, resolvedResult.RequestId);
        if (index < 0)
        {
            throw new TurnEngineInvalidRequestException(
                $"checkpoint does not contain indeterminate tool request '{resolvedResult.RequestId}'");
        }

        if (results[index].Outcome != ModelToolOutcome.Indeterminate)
        {
            throw new TurnEngineInvalidRequestException(
                $"tool request '{resolvedResult.RequestId}' is already determinate");
        }

        results[index] = resolvedResult;

        if (resolved.PendingModelResponse is null)
        {
            var messageIndex = IndexOfToolResultMessage(resolved.Messages, resolvedResult.RequestId);
            if (messageIndex < 0)
            {
                throw new TurnEngineInvalidRequestException(
                    $"checkpoint is missing the tool result message for '{resolvedResult.RequestId}'");
            }

            resolved.Messages[messageIndex] = Message.ToolResult(resolvedResult.RequestId, resolvedResult.ModelText());
        }

        resolved.ReconciliationRequired = false;

        var request = ResumeFrom(resolved, maxIterations, lastJournalSequence);
        request.ReconciliationResolution = resolvedResult;
        return request;
    }

    /// <summary>
    /// Resume after the host resolves an indeterminate model invocation, replaying the same
    /// iteration with the resolved response instead of re-invoking the model.
    /// </summary>
    /// <exception cref="TurnEngineInvalidRequestException">
    /// The checkpoint does not require model reconciliation, or the reconciliation's invocation
    /// id no longer matches the checkpoint's active invocation.
    /// </exception>
    public static TurnEngineRequest ResumeAfterModelReconciliation(
        EngineCheckpoint checkpoint,
        int maxIterations,
        long lastJournalSequence,
        ModelInvocationResponse resolvedResponse)
    {
        if (!checkpoint.ReconciliationRequired)
        {
            throw new TurnEngineInvalidRequestException("checkpoint does not require reconciliation");
        }

        var reconciliation = checkpoint.ModelReconciliation
            ?? throw new TurnEngineInvalidRequestException(
                "checkpoint requires tool reconciliation, not model reconciliation");

        if (checkpoint.ActiveInvocationId != reconciliation.InvocationId)
        {
            throw new TurnEngineInvalidRequestException(
                "model reconciliation identity does not match the active invocation");
        }

        var request = ResumeFrom(checkpoint, maxIterations, lastJournalSequence);
        request.StartIteration = checkpoint.Iteration;
        request.ReconciliationRequired = false;
        request.ModelReconciliationResolution = resolvedResponse;
        return request;
    }

    /// <summary>
    /// Build a resume request from the durable generated <see cref="ResumeContext"/>. Threads
    /// <see cref="ResumeContext.MaxModelAttempts"/> from the durable record rather than
    /// defaulting it. Use the reconciliation-specific overloads when the checkpoint is blocked
    /// pending a resolved model or tool outcome.
    /// </summary>
    public static TurnEngineRequest FromResume(ResumeContext resume)
    {
        var request = ResumeFrom(
            resume.Checkpoint,
            Math.Max(resume.MaxIterations, 0),
            Math.Max(resume.ResumeSequence(), 0));
        request.ApplyResumeAttempts(resume);
        return request;
    }

    /// <summary>
    /// Build a resume request from a <see cref="ResumeContext"/> after the host resolves an
    /// indeterminate tool effect recorded in the checkpoint.
    /// </summary>
    public static TurnEngineRequest FromResumeAfterReconciliation(ResumeContext resume, ModelToolResult resolvedResult)
    {
        var request = ResumeAfterReconciliation(
            resume.Checkpoint,
            Math.Max(resume.MaxIterations, 0),
            Math.Max(resume.ResumeSequence(), 0),
            resolvedResult);
        request.ApplyResumeAttempts(resume);
        return request;
    }

    /// <summary>
    /// Build a resume request from a <see cref="ResumeContext"/> after the host resolves an
    /// indeterminate model invocation recorded in the checkpoint.
    /// </summary>
    public static TurnEngineRequest FromResumeAfterModelReconciliation(
        ResumeContext resume,
        ModelInvocationResponse resolvedResponse)
    {
        var request = ResumeAfterModelReconciliation(
            resume.Checkpoint,
            Math.Max(resume.MaxIterations, 0),
            Math.Max(resume.ResumeSequence(), 0),
            resolvedResponse);
        request.ApplyResumeAttempts(resume);
        return request;
    }

    private void ApplyResumeAttempts(ResumeContext resume)
    {
        if (resume.MaxModelAttempts > 0)
        {
            MaxModelAttempts = resume.MaxModelAttempts;
        }
    }

    private static int IndexOfToolResult(IList<ModelToolResult> results, string requestId)
    {
        for (var i = 0; i < results.Count; i++)
        {
            if (results[i].RequestId == requestId)
            {
                return i;
            }
        }

        return -1;
    }

    private static int IndexOfToolResultMessage(IList<Message> messages, string requestId)
    {
        for (var i = 0; i < messages.Count; i++)
        {
            if (messages[i].Metadata.TryGetValue("tool_call_id", out var value) &&
                value is string toolCallId && toolCallId == requestId)
            {
                return i;
            }
        }

        return -1;
    }

    /// <summary>
    /// Shallow-clone a checkpoint's mutable collections so reconciliation patching never
    /// mutates the caller's original checkpoint instance.
    /// </summary>
    private static EngineCheckpoint CloneForPatch(EngineCheckpoint checkpoint) => new()
    {
        Id = checkpoint.Id,
        SessionId = checkpoint.SessionId,
        TurnId = checkpoint.TurnId,
        RunId = checkpoint.RunId,
        ParentRunId = checkpoint.ParentRunId,
        DelegationDepth = checkpoint.DelegationDepth,
        Iteration = checkpoint.Iteration,
        LastSequence = checkpoint.LastSequence,
        Messages = [.. checkpoint.Messages],
        StablePrefixMessages = checkpoint.StablePrefixMessages,
        Inputs = checkpoint.Inputs,
        ActiveInvocationId = checkpoint.ActiveInvocationId,
        PendingToolRequests = checkpoint.PendingToolRequests is null ? null : [.. checkpoint.PendingToolRequests],
        CompletedToolResults = checkpoint.CompletedToolResults is null ? null : [.. checkpoint.CompletedToolResults],
        CompletedModelIterations = checkpoint.CompletedModelIterations,
        ReconciliationRequired = checkpoint.ReconciliationRequired,
        ModelReconciliation = checkpoint.ModelReconciliation,
        PendingOutput = checkpoint.PendingOutput,
        FinalOutputReady = checkpoint.FinalOutputReady,
        PendingModelResponse = checkpoint.PendingModelResponse,
        ResumeSameIteration = checkpoint.ResumeSameIteration,
        PolicyAppliedForIteration = checkpoint.PolicyAppliedForIteration,
        ContextState = checkpoint.ContextState,
        Metadata = checkpoint.Metadata,
    };
}

File: runtime/java/prompty/src/main/java/com/microsoft/prompty/engine/TurnEngineRequest.java
package com.microsoft.prompty.engine;

import com.microsoft.prompty.Messages;
import com.microsoft.prompty.model.DelegatedStateReference;
import com.microsoft.prompty.model.EngineCheckpoint;
import com.microsoft.prompty.model.InvocationContextPortability;
import com.microsoft.prompty.model.Message;
import com.microsoft.prompty.model.ModelInvocationResponse;
import com.microsoft.prompty.model.ModelReconciliationState;
import com.microsoft.prompty.model.ModelToolOutcome;
import com.microsoft.prompty.model.ModelToolRequest;
import com.microsoft.prompty.model.ModelToolResult;
import com.microsoft.prompty.model.ResumeContext;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * The input to one canonical turn engine run.
 *
 * <p>This carries not just what to run but where to run from, which is what makes a turn resumable.
 * A fresh turn supplies messages and budgets; a resumed turn additionally supplies the iteration,
 * journal sequence, pending tool queue, and pending model response recovered from a checkpoint, so
 * the engine picks up exactly where the interrupted run stopped rather than repeating committed
 * effects.
 *
 * <p>Hand-written rather than generated: this is a runtime-local entry shape, not a durable
 * cross-runtime contract. The durable shapes it is built from — {@link EngineCheckpoint} and {@link
 * ResumeContext} — are generated.
 */
public final class TurnEngineRequest {

  public String sessionId = "";
  public String turnId = "";

  /** Stable identifier for this run. Empty means the engine assigns one at run start. */
  public String runId = "";

  /** The parent run this run was delegated from, or null for a top-level run. */
  public String parentRunId;

  /** Zero-based delegation nesting depth; 0 for a top-level run. */
  public int delegationDepth;

  public List<Message> messages = new ArrayList<>();
  public Object inputs;
  public int maxIterations = 10;
  public int maxModelAttempts = 3;

  /** The iteration to execute first. Non-zero when resuming a checkpoint. */
  public int startIteration;

  /** The last committed event sequence before this run. */
  public long initialSequence;

  public int stablePrefixMessages;
  public InvocationContextPortability portability = InvocationContextPortability.PORTABLE;
  public List<DelegatedStateReference> delegatedState = new ArrayList<>();
  public String activeInvocationId;
  public List<ModelToolRequest> pendingToolRequests = new ArrayList<>();
  public List<ModelToolResult> completedToolResults = new ArrayList<>();
  public int completedModelIterations;
  public boolean reconciliationRequired;
  public ModelReconciliationState modelReconciliation;
  public Object pendingOutput;
  public boolean finalOutputReady;
  public ModelInvocationResponse pendingModelResponse;
  public boolean policyAppliedForIteration;
  public ModelToolResult reconciliationResolution;
  public ModelInvocationResponse modelReconciliationResolution;

  /**
   * A fresh turn over the supplied conversation.
   *
   * <p>The whole conversation starts as the stable prefix: nothing has been appended yet, so every
   * message is eligible for provider-side prefix caching.
   */
  public static TurnEngineRequest of(String sessionId, String turnId, List<Message> messages) {
    TurnEngineRequest request = new TurnEngineRequest();
    request.sessionId = sessionId;
    request.turnId = turnId;
    request.messages = messages == null ? new ArrayList<>() : new ArrayList<>(messages);
    request.stablePrefixMessages = request.messages.size();
    return request;
  }

  /** Mark this run as delegated from a parent run, nesting one level deeper than the parent. */
  public TurnEngineRequest delegatedUnder(String parentRunId, int parentDelegationDepth) {
    this.parentRunId = parentRunId;
    this.delegationDepth =
        parentDelegationDepth == Integer.MAX_VALUE
            ? Integer.MAX_VALUE
            : parentDelegationDepth + 1;
    return this;
  }

  /** Set the stable run identifier. An empty value lets the engine assign one. */
  public TurnEngineRequest withRunId(String runId) {
    this.runId = runId;
    return this;
  }

  /**
   * Resume from a checkpoint, continuing a journal whose tail may extend past it.
   *
   * <p>Choosing the iteration is the subtle part. A checkpoint that explicitly asks to resume in
   * place does so. A checkpoint with nothing outstanding — no pending tools, no pending model
   * response, no ready output, no reconciliation — finished its iteration, so the run advances.
   * Anything else still has work inside the recorded iteration and resumes there.
   */
  public static TurnEngineRequest resumeFrom(
      EngineCheckpoint checkpoint, int maxIterations, long lastJournalSequence) {
    return resumeFrom(
        checkpoint, maxIterations, lastJournalSequence, isTrue(checkpoint.reconciliationRequired));
  }

  /**
   * The resume body, parameterised on whether reconciliation is still outstanding.
   *
   * <p>The reconciliation entry points resolve the blocking effect first and only then work out
   * where to resume, so they pass {@code false} here: a checkpoint whose sole outstanding item was
   * the effect just resolved has in fact finished its iteration, and resuming in place would rerun
   * it.
   */
  private static TurnEngineRequest resumeFrom(
      EngineCheckpoint checkpoint,
      int maxIterations,
      long lastJournalSequence,
      boolean reconciliationRequired) {
    TurnEngineRequest request = new TurnEngineRequest();
    request.sessionId = checkpoint.sessionId;
    request.turnId = checkpoint.turnId;
    request.runId = checkpoint.runId == null ? "" : checkpoint.runId;
    request.parentRunId = checkpoint.parentRunId;
    request.delegationDepth = orZero(checkpoint.delegationDepth);
    request.messages =
        checkpoint.messages == null ? new ArrayList<>() : new ArrayList<>(checkpoint.messages);
    request.stablePrefixMessages = orZero(checkpoint.stablePrefixMessages);
    request.inputs = checkpoint.inputs;
    request.maxIterations = maxIterations;
    request.maxModelAttempts = 3;

    int iteration = orZero(checkpoint.iteration);
    boolean nothingOutstanding =
        isEmpty(checkpoint.pendingToolRequests)
            && checkpoint.pendingModelResponse == null
            && !isTrue(checkpoint.finalOutputReady)
            && !reconciliationRequired;
    if (isTrue(checkpoint.resumeSameIteration)) {
      request.startIteration = iteration;
    } else if (nothingOutstanding) {
      request.startIteration = iteration + 1;
    } else {
      request.startIteration = iteration;
    }

    long checkpointSequence = checkpoint.lastSequence == null ? 0L : checkpoint.lastSequence;
    request.initialSequence = Math.max(lastJournalSequence, checkpointSequence);
    if (checkpoint.contextState != null) {
      request.portability =
          checkpoint.contextState.portability == null
              ? InvocationContextPortability.PORTABLE
              : checkpoint.contextState.portability;
      request.delegatedState =
          checkpoint.contextState.delegatedState == null
              ? new ArrayList<>()
              : new ArrayList<>(checkpoint.contextState.delegatedState);
    }
    request.activeInvocationId = checkpoint.activeInvocationId;
    request.pendingToolRequests = copyOrEmpty(checkpoint.pendingToolRequests);
    request.completedToolResults = copyOrEmpty(checkpoint.completedToolResults);
    request.completedModelIterations = orZero(checkpoint.completedModelIterations);
    request.reconciliationRequired = reconciliationRequired;
    request.modelReconciliation = checkpoint.modelReconciliation;
    request.pendingOutput = checkpoint.pendingOutput;
    request.finalOutputReady = isTrue(checkpoint.finalOutputReady);
    request.pendingModelResponse = checkpoint.pendingModelResponse;
    request.policyAppliedForIteration = isTrue(checkpoint.policyAppliedForIteration);
    return request;
  }

  /**
   * Resume after the host resolves an indeterminate tool effect.
   *
   * <p>The resolved result replaces the indeterminate one in the checkpoint, and, when the batch was
   * already folded into the conversation, the tool message the model will read is rewritten too —
   * otherwise the model would keep seeing the "outcome unknown" text the engine wrote when it gave
   * up on the effect.
   */
  public static TurnEngineRequest resumeAfterReconciliation(
      EngineCheckpoint checkpoint,
      int maxIterations,
      long lastJournalSequence,
      ModelToolResult resolvedResult) {
    if (!isTrue(checkpoint.reconciliationRequired)) {
      throw new TurnEngineException.InvalidRequest("checkpoint does not require reconciliation");
    }
    if (checkpoint.modelReconciliation != null) {
      throw new TurnEngineException.InvalidRequest(
          "checkpoint requires model reconciliation, not tool reconciliation");
    }
    if (resolvedResult.outcome == ModelToolOutcome.INDETERMINATE) {
      throw new TurnEngineException.InvalidRequest(
          "resolved tool result must have a determinate outcome");
    }

    List<ModelToolResult> results = copyOrEmpty(checkpoint.completedToolResults);
    int index = -1;
    for (int i = 0; i < results.size(); i++) {
      if (Objects.equals(results.get(i).requestId, resolvedResult.requestId)) {
        index = i;
        break;
      }
    }
    if (index < 0) {
      throw new TurnEngineException.InvalidRequest(
          "checkpoint does not contain indeterminate tool request '"
              + resolvedResult.requestId
              + "'");
    }
    if (results.get(index).outcome != ModelToolOutcome.INDETERMINATE) {
      throw new TurnEngineException.InvalidRequest(
          "tool request '" + resolvedResult.requestId + "' is already determinate");
    }
    results.set(index, resolvedResult);

    List<Message> messages = copyOrEmpty(checkpoint.messages);
    if (checkpoint.pendingModelResponse == null) {
      int messageIndex = -1;
      for (int i = 0; i < messages.size(); i++) {
        Map<String, Object> metadata = Messages.metadata(messages.get(i));
        Object toolCallId = metadata == null ? null : metadata.get(Messages.TOOL_CALL_ID);
        if (resolvedResult.requestId.equals(toolCallId)) {
          messageIndex = i;
          break;
        }
      }
      if (messageIndex < 0) {
        throw new TurnEngineException.InvalidRequest(
            "checkpoint is missing the tool result message for '"
                + resolvedResult.requestId
                + "'");
      }
      messages.set(
          messageIndex,
          Messages.toolResult(resolvedResult.requestId, ToolResults.modelText(resolvedResult)));
    }

    TurnEngineRequest request =
        resumeFrom(checkpoint, maxIterations, lastJournalSequence, false);
    request.messages = messages;
    request.completedToolResults = results;
    request.reconciliationResolution = resolvedResult;
    return request;
  }

  /** Resume after the host resolves an indeterminate model invocation. */
  public static TurnEngineRequest resumeAfterModelReconciliation(
      EngineCheckpoint checkpoint,
      int maxIterations,
      long lastJournalSequence,
      ModelInvocationResponse resolvedResponse) {
    if (!isTrue(checkpoint.reconciliationRequired)) {
      throw new TurnEngineException.InvalidRequest("checkpoint does not require reconciliation");
    }
    ModelReconciliationState reconciliation = checkpoint.modelReconciliation;
    if (reconciliation == null) {
      throw new TurnEngineException.InvalidRequest(
          "checkpoint requires tool reconciliation, not model reconciliation");
    }
    if (!Objects.equals(checkpoint.activeInvocationId, reconciliation.invocationId)) {
      throw new TurnEngineException.InvalidRequest(
          "model reconciliation identity does not match the active invocation");
    }

    TurnEngineRequest request = resumeFrom(checkpoint, maxIterations, lastJournalSequence);
    request.startIteration = orZero(checkpoint.iteration);
    request.reconciliationRequired = false;
    request.modelReconciliationResolution = resolvedResponse;
    return request;
  }

  /**
   * Resume from the durable {@link ResumeContext} record.
   *
   * <p>Unlike {@link #resumeFrom}, this threads the attempt budget from the persisted record rather
   * than defaulting it, so a resumed run keeps the retry policy the original run was given.
   */
  public static TurnEngineRequest fromResume(ResumeContext resume) {
    TurnEngineRequest request =
        resumeFrom(resume.checkpoint, maxIterationsOf(resume), resumeSequence(resume));
    applyResumeAttempts(request, resume);
    return request;
  }

  /** Resume from a {@link ResumeContext} after the host resolves an indeterminate tool effect. */
  public static TurnEngineRequest fromResumeAfterReconciliation(
      ResumeContext resume, ModelToolResult resolvedResult) {
    TurnEngineRequest request =
        resumeAfterReconciliation(
            resume.checkpoint, maxIterationsOf(resume), resumeSequence(resume), resolvedResult);
    applyResumeAttempts(request, resume);
    return request;
  }

  /** Resume from a {@link ResumeContext} after the host resolves an indeterminate invocation. */
  public static TurnEngineRequest fromResumeAfterModelReconciliation(
      ResumeContext resume, ModelInvocationResponse resolvedResponse) {
    TurnEngineRequest request =
        resumeAfterModelReconciliation(
            resume.checkpoint, maxIterationsOf(resume), resumeSequence(resume), resolvedResponse);
    applyResumeAttempts(request, resume);
    return request;
  }

  private static void applyResumeAttempts(TurnEngineRequest request, ResumeContext resume) {
    if (resume.maxModelAttempts != null && resume.maxModelAttempts > 0) {
      request.maxModelAttempts = resume.maxModelAttempts;
    }
  }

  private static int maxIterationsOf(ResumeContext resume) {
    return Math.max(resume.maxIterations == null ? 0 : resume.maxIterations, 0);
  }

  /**
   * The journal position a resumed run continues from: the further of the checkpoint's own sequence
   * and any journal tail written after it.
   */
  private static long resumeSequence(ResumeContext resume) {
    long journal = resume.lastJournalSequence == null ? 0L : resume.lastJournalSequence;
    long checkpoint =
        resume.checkpoint == null || resume.checkpoint.lastSequence == null
            ? 0L
            : resume.checkpoint.lastSequence;
    return Math.max(Math.max(journal, checkpoint), 0L);
  }

  private static boolean isTrue(Boolean value) {
    return value != null && value;
  }

  private static int orZero(Integer value) {
    return value == null ? 0 : value;
  }

  private static boolean isEmpty(List<?> list) {
    return list == null || list.isEmpty();
  }

  private static <T> List<T> copyOrEmpty(List<T> list) {
    return list == null ? new ArrayList<>() : new ArrayList<>(list);
  }
}
